File: backend/browser/click.go
package browser

import (
	"encoding/json"
	"fmt"
	"time"
)

type ClickPlan struct {
	Selector *string
	Index    int
}

type ClickProbe struct {
	X          float64 `json:"x"`
	Y          float64 `json:"y"`
	Tag        string  `json:"tag"`
	ID         string  `json:"id,omitempty"`
	ClassName  string  `json:"className,omitempty"`
	Href       string  `json:"href,omitempty"`
	Text       string  `json:"text,omitempty"`
	Selector   string  `json:"selector,omitempty"`
	Index      *int    `json:"index,omitempty"`
	MatchCount *int    `json:"matchCount,omitempty"`
}

type ClickFeedback struct {
	Clicked    bool          `json:"clicked"`
	X          float64       `json:"x"`
	Y          float64       `json:"y"`
	Tag        string        `json:"tag"`
	ID         string        `json:"id,omitempty"`
	ClassName  string        `json:"className,omitempty"`
	Href       string        `json:"href,omitempty"`
	Text       string        `json:"text,omitempty"`
	Selector   string        `json:"selector,omitempty"`
	Index      *int          `json:"index,omitempty"`
	MatchCount *int          `json:"matchCount,omitempty"`
	Download   *DownloadInfo `json:"download,omitempty"`
}

type DownloadInfo struct {
	Started  bool   `json:"started"`
	Filename string `json:"filename,omitempty"`
	URL      string `json:"url,omitempty"`
}

func buildClickProbeExpression(plan ClickPlan) (string, error) {
	clickableSelector := `a[href], button, [role="button"], input[type="submit"], input[type="button"], [onclick]`
	clickableSelectorJson, _ := json.Marshal(clickableSelector)

	if plan.Selector != nil {
		selectorJson, err := json.Marshal(*plan.Selector)
		if err != nil {
			return "", err
		}
		return fmt.Sprintf(`(() => {
			let matches;
			try {
				matches = document.querySelectorAll(%s);
			} catch (error) {
				const message = error instanceof Error ? error.message : String(error);
				throw new Error("Invalid CSS selector: " + %s + ". " + message);
			}
			const matchCount = matches.length;
			if (matchCount === 0) {
				throw new Error("Selector not found: " + %s);
			}
			const index = %d;
			if (index < 0 || index >= matchCount) {
				throw new Error("Selector " + %s + " matched " + matchCount + " element(s), but index " + index + " is out of range");
			}
			const el = matches[index];
			el.scrollIntoView({ block: "center", inline: "center" });
			const rect = el.getBoundingClientRect();
			return JSON.stringify({
				x: rect.left + rect.width / 2,
				y: rect.top + rect.height / 2,
				tag: el.tagName,
				id: el.id || undefined,
				className: typeof el.className === "string" && el.className.length > 0 ? el.className : undefined,
				href: el.href || el.getAttribute("href") || undefined,
				text: (el.textContent || "").trim().slice(0, 200) || undefined,
				selector: %s,
				index,
				matchCount,
			});
		})()`, string(selectorJson), string(selectorJson), string(selectorJson), plan.Index, string(selectorJson), string(selectorJson)), nil
	}

	return fmt.Sprintf(`(() => {
		const candidates = Array.from(document.querySelectorAll(%s));
		const index = %d;
		if (index < 0 || index >= candidates.length) {
			throw new Error("Element index " + index + " is out of range (" + candidates.length + " clickable elements)");
		}
		const el = candidates[index];
		el.scrollIntoView({ block: "center", inline: "center" });
		const rect = el.getBoundingClientRect();
		return JSON.stringify({
			x: rect.left + rect.width / 2,
			y: rect.top + rect.height / 2,
			tag: el.tagName,
			id: el.id || undefined,
			className: typeof el.className === "string" && el.className.length > 0 ? el.className : undefined,
			href: el.href || el.getAttribute("href") || undefined,
			text: (el.textContent || "").trim().slice(0, 200) || undefined,
			index,
			matchCount: candidates.length,
		});
	})()`, string(clickableSelectorJson), plan.Index), nil
}

func parseClickProbe(raw interface{}) (ClickProbe, error) {
	var probe ClickProbe
	str, ok := raw.(string)
	if ok {
		if err := json.Unmarshal([]byte(str), &probe); err != nil {
			return ClickProbe{}, err
		}
		return probe, nil
	}

	bytes, err := json.Marshal(raw)
	if err != nil {
		return ClickProbe{}, err
	}
	if err := json.Unmarshal(bytes, &probe); err != nil {
		return ClickProbe{}, err
	}
	return probe, nil
}

func dispatchMouseClick(session *CdpSession, x, y float64) error {
	type mouseParams struct {
		X          float64 `json:"x"`
		Y          float64 `json:"y"`
		Button     string  `json:"button"`
		ClickCount int     `json:"clickCount"`
		Type       string  `json:"type"`
	}

	params := mouseParams{
		X:          x,
		Y:          y,
		Button:     "left",
		ClickCount: 1,
	}

	params.Type = "mousePressed"
	_, err := session.Send("Input.dispatchMouseEvent", params)
	if err != nil {
		return err
	}

	params.Type = "mouseReleased"
	_, err = session.Send("Input.dispatchMouseEvent", params)
	return err
}

func waitForDownloadBegin(session *CdpSession, timeout time.Duration) (*DownloadInfo, error) {
	ch := make(chan *DownloadInfo, 1)

	unsubscribe := session.OnEvent("Page.downloadWillBegin", func(params interface{}) {
		var event struct {
			SuggestedFilename string `json:"suggestedFilename"`
			URL               string `json:"url"`
		}
		bytes, err := json.Marshal(params)
		if err == nil {
			_ = json.Unmarshal(bytes, &event)
		}

		ch <- &DownloadInfo{
			Started:  true,
			Filename: event.SuggestedFilename,
			URL:      event.URL,
		}
	})
	defer unsubscribe()

	select {
	case dl := <-ch:
		return dl, nil
	case <-time.After(timeout):
		return nil, nil
	}
}

func clickElementWithFeedback(session *CdpSession, plan ClickPlan) (ClickFeedback, error) {
	_, err := session.Send("Page.enable", nil)
	if err != nil {
		return ClickFeedback{}, err
	}

	expr, err := buildClickProbeExpression(plan)
	if err != nil {
		return ClickFeedback{}, err
	}

	evalVal, err := EvaluateExpression(session, expr, false)
	if err != nil {
		return ClickFeedback{}, err
	}

	probe, err := parseClickProbe(evalVal)
	if err != nil {
		return ClickFeedback{}, err
	}

	dlChan := make(chan *DownloadInfo, 1)
	go func() {
		dl, _ := waitForDownloadBegin(session, 3000*time.Millisecond)
		dlChan <- dl
	}()

	err = dispatchMouseClick(session, probe.X, probe.Y)
	if err != nil {
		return ClickFeedback{}, err
	}

	dl := <-dlChan

	feedback := ClickFeedback{
		Clicked:    true,
		X:          probe.X,
		Y:          probe.Y,
		Tag:        probe.Tag,
		ID:         probe.ID,
		ClassName:  probe.ClassName,
		Href:       probe.Href,
		Text:       probe.Text,
		Selector:   probe.Selector,
		Index:      probe.Index,
		MatchCount: probe.MatchCount,
		Download:   dl,
	}
	return feedback, nil
}

File: backend/browser/click.ts
// PORT: backend/browser/click.go

import { Effect } from "effect";
import { type CdpSession, EvaluateExpression } from "./cdp";

export interface ClickPlan {
  Selector: string | null;
  Index: number;
}

export interface ClickProbe {
  x: number;
  y: number;
  tag: string;
  id?: string;
  className?: string;
  href?: string;
  text?: string;
  selector?: string;
  index?: number;
  matchCount?: number;
}

export interface DownloadInfo {
  started: boolean;
  filename?: string;
  url?: string;
}

export interface ClickFeedback {
  clicked: boolean;
  x: number;
  y: number;
  tag: string;
  id?: string;
  className?: string;
  href?: string;
  text?: string;
  selector?: string;
  index?: number;
  matchCount?: number;
  download?: DownloadInfo;
}

export function buildClickProbeExpression(plan: ClickPlan): Effect.Effect<string, Error> {
  const clickableSelector = `a[href], button, [role="button"], input[type="submit"], input[type="button"], [onclick]`;
  const clickableSelectorJson = JSON.stringify(clickableSelector);

  if (plan.Selector !== null) {
    let selectorJson: string;
    try {
      selectorJson = JSON.stringify(plan.Selector);
    } catch (err) {
      return Effect.fail(err instanceof Error ? err : new Error(String(err)));
    }
    const expr = `(() => {
			let matches;
			try {
				matches = document.querySelectorAll(${selectorJson});
			} catch (error) {
				const message = error instanceof Error ? error.message : String(error);
				throw new Error("Invalid CSS selector: " + ${selectorJson} + ". " + message);
			}
			const matchCount = matches.length;
			if (matchCount === 0) {
				throw new Error("Selector not found: " + ${selectorJson});
			}
			const index = ${plan.Index};
			if (index < 0 || index >= matchCount) {
				throw new Error("Selector " + ${selectorJson} + " matched " + matchCount + " element(s), but index " + index + " is out of range");
			}
			const el = matches[index];
			el.scrollIntoView({ block: "center", inline: "center" });
			const rect = el.getBoundingClientRect();
			return JSON.stringify({
				x: rect.left + rect.width / 2,
				y: rect.top + rect.height / 2,
				tag: el.tagName,
				id: el.id || undefined,
				className: typeof el.className === "string" && el.className.length > 0 ? el.className : undefined,
				href: el.href || el.getAttribute("href") || undefined,
				text: (el.textContent || "").trim().slice(0, 200) || undefined,
				selector: ${selectorJson},
				index,
				matchCount,
			});
		})()`;
    return Effect.succeed(expr);
  }

  const expr = `(() => {
		const candidates = Array.from(document.querySelectorAll(${clickableSelectorJson}));
		const index = ${plan.Index};
		if (index < 0 || index >= candidates.length) {
			throw new Error("Element index " + index + " is out of range (" + candidates.length + " clickable elements)");
		}
		const el = candidates[index];
		el.scrollIntoView({ block: "center", inline: "center" });
		const rect = el.getBoundingClientRect();
		return JSON.stringify({
			x: rect.left + rect.width / 2,
			y: rect.top + rect.height / 2,
			tag: el.tagName,
			id: el.id || undefined,
			className: typeof el.className === "string" && el.className.length > 0 ? el.className : undefined,
			href: el.href || el.getAttribute("href") || undefined,
			text: (el.textContent || "").trim().slice(0, 200) || undefined,
			index,
			matchCount: candidates.length,
		});
	})()`;
  return Effect.succeed(expr);
}

export function parseClickProbe(raw: any): Effect.Effect<ClickProbe, Error> {
  return Effect.try({
    try: () => {
      if (typeof raw === "string") {
        return JSON.parse(raw) as ClickProbe;
      }
      if (typeof raw === "object" && raw !== null) {
        return {
          x: Number(raw.x),
          y: Number(raw.y),
          tag: String(raw.tag),
          id: raw.id !== undefined ? String(raw.id) : undefined,
          className: raw.className !== undefined ? String(raw.className) : undefined,
          href: raw.href !== undefined ? String(raw.href) : undefined,
          text: raw.text !== undefined ? String(raw.text) : undefined,
          selector: raw.selector !== undefined ? String(raw.selector) : undefined,
          index: raw.index !== undefined ? Number(raw.index) : undefined,
          matchCount: raw.matchCount !== undefined ? Number(raw.matchCount) : undefined,
        };
      }
      throw new Error(`Invalid raw click probe type: ${typeof raw}`);
    },
    catch: (cause) => (cause instanceof Error ? cause : new Error(String(cause))),
  });
}

export function dispatchMouseClick(
  session: CdpSession,
  x: number,
  y: number,
): Effect.Effect<void, Error> {
  const baseParams = {
    x,
    y,
    button: "left",
    clickCount: 1,
  };

  return session.Send("Input.dispatchMouseEvent", { ...baseParams, type: "mousePressed" }).pipe(
    Effect.flatMap(() =>
      session.Send("Input.dispatchMouseEvent", { ...baseParams, type: "mouseReleased" })
    ),
    Effect.map(() => undefined)
  );
}

export function waitForDownloadBegin(
  session: CdpSession,
  timeoutMs: number,
): Effect.Effect<DownloadInfo | null, Error> {
  return Effect.async<DownloadInfo | null, Error>((resume) => {
    let completed = false;

    const unsubscribe = session.OnEvent("Page.downloadWillBegin", (params: any) => {
      if (completed) return;
      completed = true;
      clearTimeout(timer);
      unsubscribe();

      const suggestedFilename = params?.suggestedFilename || "";
      const url = params?.url || "";

      resume(
        Effect.succeed({
          started: true,
          filename: suggestedFilename,
          url: url,
        })
      );
    });

    const timer = setTimeout(() => {
      if (completed) return;
      completed = true;
      unsubscribe();
      resume(Effect.succeed(null));
    }, timeoutMs);
  });
}

export function clickElementWithFeedback(
  session: CdpSession,
  plan: ClickPlan,
): Effect.Effect<ClickFeedback, Error> {
  return session.Send("Page.enable", null).pipe(
    Effect.flatMap(() => buildClickProbeExpression(plan)),
    Effect.flatMap((expr) => EvaluateExpression(session, expr, false)),
    Effect.flatMap((evalVal) => parseClickProbe(evalVal)),
    Effect.flatMap((probe) =>
      Effect.all([
        waitForDownloadBegin(session, 3000),
        dispatchMouseClick(session, probe.x, probe.y),
      ], { concurrency: "unbounded" }).pipe(
        Effect.map(([dl, _]) => {
          const feedback: ClickFeedback = {
            clicked: true,
            x: probe.x,
            y: probe.y,
            tag: probe.tag,
            id: probe.id,
            className: probe.className,
            href: probe.href,
            text: probe.text,
            selector: probe.selector,
            index: probe.index,
            matchCount: probe.matchCount,
          };
          if (dl !== null) {
            feedback.download = dl;
          }
          return feedback;
        })
      )
    )
  );
}

/*
PORT STATUS
source path: backend/browser/click.go
source lines: 240
draft lines: 211
confidence: high
status: phase_b_compile
*/
